pkg/controller/core.oam.dev/v1alpha2/application/dispatcher.go

```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+// manifestDispatcher is a manifest dispatcher
+type manifestDispatcher struct {
+	run         func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision, clusterName string) (bool, error)
+	healthCheck func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision) (bool, error)
+}
+
+func (h *AppHandler) generateDispatcher(appRev *v1beta1.ApplicationRevision, readyWorkload *unstructured.Unstructured, readyTraits []*unstructured.Unstructured, overrideNamespace string) ([]*manifestDispatcher, error) {
+	dispatcherGenerator := func(options DispatchOptions) *manifestDispatcher {
+		assembleManifestFn := func(skipApplyWorkload bool) (bool, []*unstructured.Unstructured) {
+			manifests := options.Traits
+			skipWorkload := skipApplyWorkload || options.Workload == nil
+			if !skipWorkload {
+				manifests = append([]*unstructured.Unstructured{options.Workload}, options.Traits...)
+			}
+			return skipWorkload, manifests
+		}
+
+		dispatcher := new(manifestDispatcher)
+		dispatcher.healthCheck = func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision) (bool, error) {
+			skipWorkload, manifests := assembleManifestFn(wl.SkipApplyWorkload)
+			if !h.resourceKeeper.ContainsResources(manifests) {
+				return false, nil
+			}
+			_, isHealth, err := h.collectHealthStatus(ctx, wl, appRev, options.OverrideNamespace, skipWorkload,
+				ByTraitType(readyTraits, options.Traits))
+			if err != nil {
+				return false, err
+			}
+			return isHealth, nil
+		}
+		dispatcher.run = func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision, clusterName string) (bool, error) {
+			skipWorkload, dispatchManifests := assembleManifestFn(wl.SkipApplyWorkload)
+			if isHealth, err := dispatcher.healthCheck(ctx, wl, appRev); !isHealth || err != nil {
+				if err := h.Dispatch(ctx, clusterName, common.WorkflowResourceCreator, dispatchManifests...); err != nil {
+					return false, errors.WithMessage(err, "Dispatch")
+				}
+				status, isHealth, err := h.collectHealthStatus(ctx, wl, appRev, options.OverrideNamespace, skipWorkload,
+					ByTraitType(readyTraits, options.Traits))
+				if err != nil {
+					return false, errors.WithMessage(err, "CollectHealthStatus")
+				}
+				if options.Stage < DefaultDispatch {
+					status.Healthy = false
+					if status.Message == "" {
+						status.Message = "waiting for previous stage healthy"
+					}
+					h.addServiceStatus(true, *status)
+				}
+				if !isHealth {
+					return false, nil
+				}
+			}
+			return true, nil
+		}
+		return dispatcher
+	}
+
+	traitStageMap := make(map[StageType][]*unstructured.Unstructured)
+	for _, readyTrait := range readyTraits {
+		var (
+			traitType = readyTrait.GetLabels()[oam.TraitTypeLabel]
+			stageType = DefaultDispatch
+			err       error
+		)
+		switch {
+		case traitType == definition.AuxiliaryWorkload:
+		case traitType != "":
+			if strings.Contains(traitType, "-") {
```

Comment:
why do we need this check logic?

Reply:
can you give some concrete examples?

Reply:
I got it, but why it's not handled outside? AFAIK, it should be trimed when getting the traitdefition spec detial before the application rendered.